File: ios/Store/ImageStore.swift
import Foundation

struct ImageStoreResult {

    let id: String

    let cacheFile: URL

    let permanentFile: URL

}

enum ImageStore {

    static func prepare(
        format: ImageFormat
    ) throws -> ImageStoreResult {

        let id = ImageIdGenerator.generate()

        let ext: String

        switch format {

        case .jpeg:
            ext = "jpg"

        case .png:
            ext = "png"

        case .webp:
            ext = "webp"
        }

        return ImageStoreResult(

            id: id,

            cacheFile: try CacheManager.createFile(
                id: id,
                extension: ext
            ),

            permanentFile: try SpiralFileManager.createResizeFile(
                id: id,
                extension: ext
            )
        )
    }

    static func persist(
        _ result: ImageStoreResult
    ) throws {

        if FileManager.default.fileExists(
            atPath: result.permanentFile.path
        ) {

            try FileManager.default.removeItem(
                at: result.permanentFile
            )

        }

        try FileManager.default.copyItem(
            at: result.cacheFile,
            to: result.permanentFile
        )

    }

}

File: android/src/main/java/com/margelo/nitro/spiralimage/store/ImageStore.kt
package com.margelo.nitro.spiralimage.store

import android.content.Context
import com.margelo.nitro.spiralimage.ImageFormat
import com.margelo.nitro.spiralimage.ProcessOptions
import java.io.File

data class ImageStoreResult(
    val id: String,
    val cacheFile: File,
    val permanentFile: File
)

object ImageStore {

    fun prepare(
        context: Context,
        options: ProcessOptions
    ): ImageStoreResult {

        val id = ImageIdGenerator.generate(options.path)

        val format = options.output?.format ?: ImageFormat.JPEG

        val extension = when (format) {
            ImageFormat.JPEG -> "jpg"
            ImageFormat.PNG -> "png"
            ImageFormat.WEBP -> "webp"
        }

        return ImageStoreResult(
            id = id,

            cacheFile = CacheManager.createFile(
                cacheDir = context.cacheDir,
                imageId = id,
                extension = extension
            ),

            permanentFile = FileManager.createResizeFile(
                context = context,
                imageId = id,
                extension = extension
            )
        )
    }

    fun persist(
        result: ImageStoreResult
    ) {
        result.cacheFile.copyTo(
            target = result.permanentFile,
            overwrite = true
        )
    }
}